server/src/services/asset-media.service.ts

```diff
@@ -57,7 +70,118 @@ export class AssetMediaService {
     this.access = AccessCore.create(accessRepository);
   }
 
-  public async replaceAsset(
+  async getUploadAssetIdByChecksum(auth: AuthDto, checksum?: string): Promise<AssetMediaResponseDto | undefined> {
+    if (!checksum) {
+      return;
+    }
+
+    const assetId = await this.assetRepository.getUploadAssetIdByChecksum(auth.user.id, fromChecksum(checksum));
+    if (!assetId) {
+      return;
+    }
+
+    return { id: assetId, status: AssetMediaStatus.DUPLICATE };
+  }
+
+  canUploadFile({ auth, fieldName, file }: UploadRequest): true {
+    this.access.requireUploadAccess(auth);
+
+    const filename = file.originalName;
+
+    switch (fieldName) {
+      case UploadFieldName.ASSET_DATA: {
+        if (mimeTypes.isAsset(filename)) {
+          return true;
+        }
+        break;
+      }
+
+      case UploadFieldName.SIDECAR_DATA: {
+        if (mimeTypes.isSidecar(filename)) {
+          return true;
+        }
+        break;
+      }
+
+      case UploadFieldName.PROFILE_DATA: {
+        if (mimeTypes.isProfile(filename)) {
+          return true;
+        }
+        break;
+      }
+    }
+
+    this.logger.error(`Unsupported file type ${filename}`);
+    throw new BadRequestException(`Unsupported file type ${filename}`);
+  }
+
+  getUploadFilename({ auth, fieldName, file }: UploadRequest): string {
+    this.access.requireUploadAccess(auth);
+
+    const originalExtension = extname(file.originalName);
+
+    const lookup = {
+      [UploadFieldName.ASSET_DATA]: originalExtension,
+      [UploadFieldName.SIDECAR_DATA]: '.xmp',
+      [UploadFieldName.PROFILE_DATA]: originalExtension,
+    };
+
+    return sanitize(`${file.uuid}${lookup[fieldName]}`);
+  }
+
+  getUploadFolder({ auth, fieldName, file }: UploadRequest): string {
+    auth = this.access.requireUploadAccess(auth);
+
+    let folder = StorageCore.getNestedFolder(StorageFolder.UPLOAD, auth.user.id, file.uuid);
+    if (fieldName === UploadFieldName.PROFILE_DATA) {
+      folder = StorageCore.getFolderLocation(StorageFolder.PROFILE, auth.user.id);
+    }
+
+    this.storageRepository.mkdirSync(folder);
+
+    return folder;
+  }
+
+  async uploadFile(
+    auth: AuthDto,
+    dto: AssetMediaCreateDto,
+    file: UploadFile,
+    sidecarFile?: UploadFile,
+  ): Promise<AssetMediaResponseDto> {
+    try {
+      await this.access.requirePermission(
+        auth,
+        Permission.ASSET_UPLOAD,
+        // do not need an id here, but the interface requires it
+        auth.user.id,
+      );
+
+      if (dto.livePhotoVideoId) {
```

Comment:
This code doesn't need to be here, as livePhotos should now be uploaded in their own request, and linked with a PUT request on the asset record.

Reply:
Why not just link on create?

Reply:
Oh right, you could upload the live photo first, get its id, and then create the primary photo and link it at the same time. Ok, this works. However, I think that means the the modifications to the dart client are backwards. If i'm not mistaken, it appears that the primary photo id is being set as the live photo of the of the live photo.

Reply:
I think it should be possible to upload the primary photo first, and then upload the live photo, and then link the two, which requires a change in asset.service's PUT method to allow the new property to be edited.

Reply:
This is a good idea. We can add this in a follow-up PR.